### PR TITLE
Ttest greta

### PR DESCRIPTION
**Pull request by gretaabib**
Hi, 

As requested in the PR: https://github.com/tnc-br/ddf-isoscapes/pull/152

raster.py -> I’ve changed the functions coords_to_indices and get_data_at_coords
dataset.py -> I've included the create_fraudulent_samples function and its dependencies.

### CODE REVIEW

File: __pycache__/dataset.cpython-310.pyc
Comment (erickzul) [inline __pycache__/dataset.cpython-310.pyc:1]: Could you delete this and the rest of the .pyc files?

File: dataset.py [hunk 1]
```diff
@@ -137,3 +142,91 @@ def yearly_255_trees_partitioned(reference_csv_filename: str) -> PartitionedData
   print_split(partition_data)
   return partition_data
 
+#Utility function for randomly sampling a point around a sample site
+def is_valid_point(lat: float, lon: float, reference_isocape: raster.AmazonGeoTiff):
+  return True if raster.get_data_at_coords(reference_isocape, lon, lat, 0) else False
+
+# Pick a random point around (lat, lon) within max_distance_km. If edge_only is
+# true, only pick points exactly max_distance_km away from (lat, lon).
+def random_nearby_point(lat, lon, max_distance_km, edge_only=False):
```
Comment (erickzul) [inline dataset.py:151]: Could we specify the parameter data types of the function like you do for is_valid_point?

File: dataset.py [hunk 2]
```diff
@@ -137,3 +142,91 @@ def yearly_255_trees_partitioned(reference_csv_filename: str) -> PartitionedData
...
+# of any of those points.
+def is_nearby_real_point(lat, lon, real_points, threshold_km):
+  for point, _ in real_points:
+    if distance.geodesic((lat, lon), point).km < threshold_km:
+      return True
+  return False
+
+
+#This function creates a dataset based on real samples adding a Fraud column
+def create_fraudulent_samples(real_samples_data, mean_iso,element,max_trusted_radius,max_fraud_radius,min_fraud_radius):
```
Comment (erickzul) [inline dataset.py:176]: From the functions you added, I think this is the only one we want to surface as a public API. Could we replace the "helper" functions to have an underscore such that we follow the standard to name them like that if we don't intend them to use them beyond being helper functions? i.e. _is_nearby_real_point

Comment (erickzul) [reply dataset.py:176]: Also, please add function parameter data types.

File: dataset.py [hunk 3]
```diff
@@ -137,3 +142,91 @@ def yearly_255_trees_partitioned(reference_csv_filename: str) -> PartitionedData
...
+        max_fraud_radius, min_fraud_radius))
+    
+  fake_sample = pd.DataFrame(columns=['Code',
+          'lat',
+          'long',
+          element,
+          'fraud'])
+
+  # Max number of times to attempt to generate random coordinates.
+  MAX_RANDOM_SAMPLE_ATTEMPTS = 1000
```
Comment (erickzul) [inline dataset.py:210]: Given that this is not a global constant, could we change this to be lowercase?

File: dataset.py [hunk 4]
```diff
@@ -137,3 +142,91 @@ def yearly_255_trees_partitioned(reference_csv_filename: str) -> PartitionedData
...
+  '''
+
+  real_samples_data.dropna(subset=[element], how='all', inplace=True)
+  real_samples = real_samples_data.groupby(['lat','long'])[element]
+  real_samples_code = real_samples_data.groupby(['lat','long','Code'])[element]
+
+  count = 0
+  lab_samp = real_samples
+  #start number for fake code
+  fake_code_start = 5000
```
Comment (erickzul) [inline dataset.py:197]: Is this guaranteed to be unique i.e. can we say there will never be a mad5000 in our data, especially if we expect to get more data in the future?

I'd suggest changing Code formulation below to be unique to fraudulent samples i.e. f"fake_mad{fake_code_start+count}.

File: dataset.py [hunk 5]
```diff
@@ -137,3 +142,91 @@ def yearly_255_trees_partitioned(reference_csv_filename: str) -> PartitionedData
...
+    while((not is_valid_point(lat, lon, mean_iso) or
+          is_nearby_real_point(lat, lon, real_samples, min_fraud_radius)) and
+          attempts < MAX_RANDOM_SAMPLE_ATTEMPTS):
+      lat, lon = random_nearby_point(coord[0], coord[1], max_fraud_radius)
+      if lab_samp.size < 5:
+        pass
+      else:
+      #generating 5 samples per code
+        for i in range(5):
+          new_row = {'Code': 'mad'+str(fake_code_start+count), 'lat': lat, 'long': lon,element: lab_samp.iloc[i],'fraud': True }
```
Comment (erickzul) [inline dataset.py:227]: Please replace 'mad'+str(fake_code_start+count) with f"fake{(fake_code_start+count)}

File: dataset.py [hunk 6]
```diff
@@ -137,3 +142,91 @@ def yearly_255_trees_partitioned(reference_csv_filename: str) -> PartitionedData
...
+          fake_sample.loc[len(fake_sample)] = new_row
+      attempts += 1
+
+  return(fake_sample)
```
Comment (erickzul) [inline dataset.py:231]: Please replace with "return fake_sample". No parenthesis are needed.

File: hypothesis.py [hunk 7]
```diff
@@ -89,6 +88,7 @@ def fraud_metrics(sample_data: pd.DataFrame,
         _LONGITUDE_COLUMN_NAME,
         _LATITUDE_COLUMN_NAME,
         _FRAUDULENT_COLUMN_NAME])[isotope_column_name]
+
```
Comment (erickzul) [inline hypothesis.py:91]: Could we revert the changes on this file? It looks like we're only adding line breaks that we don't need for this change.

File: raster.py [hunk 8]
```diff
@@ -223,8 +223,9 @@ def save_numpy_to_geotiff(bounds: Bounds, prediction: np.ma.MaskedArray, path: s
 
 def coords_to_indices(bounds: Bounds, x: float, y: float):
   if x < bounds.minx or x > bounds.maxx or y < bounds.miny or y > bounds.maxy:
-    raise ValueError("Coordinates out of bounds")
-
+    #raise ValueError("Coordinates out of bounds")
```
Comment (erickzul) [inline raster.py:226]: Why do we want to do this? Do we not want to make our program raise this anomaly?

File: dataset.py [hunk 9]
```diff
@@ -137,3 +142,91 @@ def yearly_255_trees_partitioned(reference_csv_filename: str) -> PartitionedData
   print_split(partition_data)
   return partition_data
 
+#Utility function for randomly sampling a point around a sample site
+def is_valid_point(lat: float, lon: float, reference_isocape: raster.AmazonGeoTiff):
+  return True if raster.get_data_at_coords(reference_isocape, lon, lat, 0) else False
```
Comment (erickzul) [inline dataset.py:147]: If the reason to change get_data_at_cords to not raise the exception, there are other ways to do this without changing it i.e.

try:
   _ = raster.get_data_at_coords(reference_isoscape, lon, lat, 0)
  return True
except ValueError:
   return False

**Review by erickzul — CHANGES_REQUESTED** [review]

File: raster.py (hunk 8 above)
Comment (gretaabib) [reply raster.py:226]: It was a change made by Ruben in order to generate the random coordinates. If we keep as is, we got "coordinates out of bounds" all the time =/

File: __pycache__/dataset.cpython-310.pyc
Comment (gretaabib) [reply __pycache__/dataset.cpython-310.pyc:1]: Done

File: dataset.py (hunk 5 above)
Comment (gretaabib) [reply dataset.py:227]: Code pattern received from Dr. Martinelli's lab is madxxx

File: dataset.py (hunk 2 above)
Comment (erickzul) [reply dataset.py:176]: Per offline conversation, please add dataset_test.ipynb where we test only create_fraudulent_samples.

**Review by erickzul — CHANGES_REQUESTED** [review]

File: dataset.py (hunk 5 above)
Comment (gretaabib) [reply dataset.py:227]: As we don't need the fake_code_start, it was removed,
'Code': f"fake_mad{count}"

File: dataset.py (hunk 4 above)
Comment (gretaabib) [reply dataset.py:197]: 'Code': f"fake_mad{count}"

File: dataset.py (hunk 2 above)
Comment (rothn) [inline dataset.py:176]: Consider documenting some of these parameters, especially `element`

**Review by rothn — CHANGES_REQUESTED** [review]
If you need to pass on the suggested changes, feel free to justify and do so. Otherwise, the fixes should be relatively minor + clarifications except for the sampling strategy

File: dataset.py [hunk 10]
```diff
@@ -137,3 +142,91 @@ def yearly_255_trees_partitioned(reference_csv_filename: str) -> PartitionedData
...
+  # Pick a random angle pointing outward from the origin.
+  # 0 == North, 90 == East, 180 == South, 270 == West
+  angle = 360 * random.random()
+
+  # sqrt() is required for an equal radial distribution, otherwise samples
+  # cluster around origin.
+  dist = max_distance_km if edge_only else max_distance_km * math.sqrt(random.random())
+
+  # WGS-84 is the most accurate ellipsoidal model of Earth, but we should double
+  # check to make sure this matches the model used by our sample collectors.
```
[the remark this answers is not in the document]
Comment (gretaabib) [reply dataset.py:161]: @VimWizardVersace could you please help me on this?

**Comment by rothn** [comment]
@gretaabib  FYI the .pyc files are back. Have you considered adding *.pyc to the .gitignore file?

**Review by rothn — APPROVED** [review]

**Comment by gretaabib** [comment]
@erickzul please review [dataset_test.ipynb](https://github.com/tnc-br/ddf_common/blob/ttest-greta/dataset_test.ipynb)

**Review by erickzul — DISMISSED** [review]
Please update dataset.py with the latest changes. It's missing the preprocess_sample_data function.

**Review by gretaabib — COMMENTED** [review]
Checking the conflicts

**Review by erickzul — APPROVED** [review]
We may want to revisit returning in coords to indices as the exception may be important to consider instead of None. For now, LGTM.
